### PR TITLE
Implement player events list scrolling behavior

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Builds off of this PR #6971 as part of the new recording playback revamp.

This PR more tightly couples the event list with the playback controller/frame and implements autoscrolling of the events list. 

This is now possible:

//V1: https://user-images.githubusercontent.com/13460330/141377760-77a7f3ae-d66b-49ff-9e1a-400331f644c7.mov
//V2:

https://user-images.githubusercontent.com/13460330/142083795-581296b7-02b1-4ba8-872e-77239be15218.mov

### Some areas that can be optimized:

- [x] ~~Autoscrolling faster than we can render~~ Workaround this with new "jump to current" behavior

//V1: https://user-images.githubusercontent.com/13460330/141376287-cf289b57-61ad-4876-a7e2-814a1caa379c.mov

Sunk too much time into this one so would like some guidance here. The video above shows what happens when we try to auto scroll faster than the DOM can render rows. This happens because virtualized lists make the tradeoff of rendering only the rows in the viewframe of the container. When a user scrolls away from the visible part of the list, the virtualized list will determine the new list of events to render using the new scroll position. This gives us a huge performance optimization when we're dealing with long lists of events, but leaves us with the dilemma of ugly loading states when we're autoscrolling too fast. The current workaround (that doesn't always when the scroll jump is too big) is to use [overscanning](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#prop-types) which renders a specific number of rows (set to 50 right now) above and below the ones that are visible.

- [ ] Scrollbar jumping when dealing with long recordings with many events

We're fetching events in chunks, and we render them in the events list lazily upon arrival. This also means that `react-virtualized` autosizer has to measure the rows heights all over again on every `loadEventsSuccess` action which can get pretty expensive. This also has weird side effects such as the scrollbar jumping as more chunks get loaded in while the player is playing, or scrollbar drag not working/being delayed.

- [ ] Row height sizing isn't always accurate (and making it accurate could be expensive)

I think this might be of a shortcoming of `react-virtualized` but row height sizes aren't always accurate. Because row contains dynamic content, we measure the size of each cell to calculate the parent container's total height. Resizing the window a bit might knock these into place, but `clearRowCache` is an expensive process if we're dealing with a lot of rows.

## How did you test this code?

Add logic tests for any new functionality in `eventListLogic`

### CODE REVIEW

**Review by clarkus — DISMISSED** [review]
This is so cool! My one concern aligns with your high speed playback example. That might be triggering for users who react physically to flashing or flickering images (https://developer.mozilla.org/en-US/docs/Web/Accessibility/Seizure_disorders). I wonder if we might show some placeholder spinner when this is happening? At rapid playback speeds that event list is going to be pretty useless. @paolodamico any thoughts on what to do here? 

We might also consider pausing autoplay when the user has otherwise modified that list offset. Say you hover over that and scroll up to view a different event. We wouldn't want to autoplay the user out of that context even if the recording was still playing.

**Comment by alexkim205** [comment]
I can see the stop autoscroll on hover over list behavior being super intuitive to use. 

As for the rapid playback speeds causing the events list to flicker, maybe we can disable the autoscroll feature for speeds greater than 4x? Something that could also help with list jittering would be to disable autoscrolling until all the events and snapshots are loaded. However, I'm not entirely sure what the best way to expose this to the user is.

**Comment by clarkus** [comment]
Another idea might be paging the event list based on the playback speed. For example, only moving through in sets instead of rapidly for each item. That would mean less redraws, less movement, but still a representative playback experience.

**Comment by paolodamico** [comment]
My perspective:

1. I think intuitively if you're hovering over the list, the list should stop scrolling automatically but playback should continue, you may be looking for an event but still want to keep watching the recording.
2. For large playback speeds, I agree with Chris on moving by sections, and maybe we highlight a section instead of a specific event?

**Comment by clarkus** [comment]
Another idea from this morning - a jump-to action that lets you catch up to the current events. Instead of autoscrolling, we auto-highlight and make it easy to jump to that context.

**Comment by alexkim205** [comment]
Thanks everyone for the feedback. I'll briefly summarize some of the points brought up here and offline.

- [x] Remove autoscrolling
- [x] Implement catchup mechanism to scroll to current event. Show action button whenever user scrolls current event out of frame.

Screen Shot 2021-11-12 at 10 56 59 AM

- [x] Highlight sections of events with same timestamp, not specific events.
- [x] Don't show events list until they're all loaded? Maybe same applies for seekbar events.
- [x] Animate highlighted events
- ~~Stop autoscrolling on hover over list~~ - Irrelevant now that we're removing autoscroll
- ~~Paging events list~~ - Irrelevant now that we're removing autoscroll

**Comment by alexkim205** [comment]
Hi all, retagging @rcmarron @clarkus since this PR is ready for review again. A few important things have changed since the first version, namely the way we keep the current events in frame. 

This revision also fixes a really annoying performance issue where entire DOM was rerendering when rrweb replays each snapshot. This turned out to be related to how often `react-virtualized` renders its cells and was fixed by overwriting rerender pattern with `useCallback`. Expensive repaints were also causing scrolling/jump to scrolling to be painfully slow so scrolling should be a bit more tolerable.

Demo:

https://user-images.githubusercontent.com/13460330/142083795-581296b7-02b1-4ba8-872e-77239be15218.mov

Edge detection:

https://user-images.githubusercontent.com/13460330/142085181-b5f0428c-1c14-4b49-a699-d460da525608.mov

Some nuances to consider and maybe iterate on in later PRs:

- When you try to jump to the current event while the player is playing through multiple events, you'll be taken to a stale scroll position that is no longer highlighted. This can be somewhat confusing for users who expect "jump to current" to mean "jump to current and stick with current" (kind of how Slack operates). Since we're not going to implement the "stick to current" functionality, I wonder if there's a better way to communicate this perceived lag. @clarkus 
- Highlight box height isn't always accurate. Same problem as "Row height sizing isn't always accurate (and making it accurate could be expensive)" in PR description.
- Added an X button to the position finder. As of now clicking the x will mean that the position finder won't resurface until you reload the entire page. In the case after a user closes out the finder, it might make sense to allow the finder to resurface again when the view frame overlaps with the current events frame. However I don't know if its worth the dev time to implement this.

**Comment by clarkus** [comment]
> When you try to jump to the current event while the player is playing through multiple events, you'll be taken to a stale scroll position that is no longer highlighted. This can be somewhat confusing for users who expect "jump to current" to mean "jump to current and stick with current" (kind of how Slack operates). Since we're not going to implement the "stick to current" functionality, I wonder if there's a better way to communicate this perceived lag. @clarkus

How about "jump to recent" or "jump to current time" - we're really putting them in the context of a timestamp - there could be multiple events. I don't think we need to be too targeted with our language - just something that communicates you can jump to the stuff you just saw play

**Comment by clarkus** [comment]
> Highlight box height isn't always accurate. Same problem as "Row height sizing isn't always accurate (and making it accurate could be expensive)" in PR description.

Auto-capture feels like the exception here. It needs extra description to be meaningful and there is a very clear primary attribute for describing the event. We can either make this case more defined, or find a way to describe other events in similarly descriptive ways. For example, current URL is a great identifier for a pageview event. I think focusing on auto-capture makes sense for now, but eventually we should be able to get to a more predictable pattern for these items.

**Comment by alexkim205** [comment]
> I think focusing on auto-capture makes sense for now, but eventually we should be able to get to a more predictable pattern for these items.

From the implementation side, this would be amazing and so much quicker to render. Knowing the row heights for each event type beforehand (i.e., Autocapture -> 3 lines, pageevents -> 2 lines, etc.) would mean `react-virtualized` can precalculate the total height without `<CellMeasurer/>` bloat. I'll refactor some of this so that we can add/remove/change any events we see fit into the an eventsToRowHeights map

**Comment by clarkus** [comment]
> From the implementation side, this would be amazing and so much quicker to render. Knowing the row heights for each event type beforehand (i.e., Autocapture -> 3 lines, pageevents -> 2 lines, etc.) would mean `react-virtualized` can precalculate the total height without `<CellMeasurer/>` bloat. I'll refactor some of this so that we can add/remove/change any events we see fit into the an eventsToRowHeights map

Brought up some examples in figma to help work through this. https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf?node-id=5186:36628#125681661

**Review by rcmarron — DISMISSED** [review]
Looking great - I tried to find bugs but it was hard! A couple small things:
* We might want a bit of margin at the top of the list. The blue highlight goes right into the search box.
image
* The event loading indicator never seems to appear.
* Do we want to handle the "no events" case. So showing some message like "No events fired in this recording".

File: frontend/src/scenes/session-recordings/player/PlayerEvents.tsx [hunk 1]
```diff
@@ -1,66 +1,144 @@
...
+                        className={clsx('event-list-item', { 'current-event': isEventCurrent(index) })}
+                        align="top"
+                        style={style}
+                    >
+                        <Col className="event-item-icon">
+                            <div className="event-item-icon-wrapper">{renderIcon(event)}</div>
+                        </Col>
+                        <Col
+                            className={clsx('event-item-content', {
+                                rendering: !renderAllRows && !isRowIndexRendered(index),
```
[the remark this answers is not in the document]
Comment (alexkim205) [reply frontend/src/scenes/session-recordings/player/PlayerEvents.tsx:92]: good catch, looks like i overthought this one.

**Comment by alexkim205** [comment]
Thanks for the reviews everyone ❤️ . I'm going to offload the row heights work into a smaller PR to get this one in quickly. @rcmarron addressed your feedback! Now events shouldn't show up in the list until all the events are loaded. A loading indicator takes its place and this also helps with any scroll jumping that we had before. Empty and loading states are also properly separated out.

**Review by rcmarron — APPROVED** [review]
LGTM!

One last little thing. It looks like the events list is staying in loading until the entire recording is loaded - not just the events.

**Comment by alexkim205** [comment]
> LGTM!
> 
> One last little thing. It looks like the events list is staying in loading until the entire recording is loaded - not just the events.

Hm I did that to prevent the empty data state from showing and flickering to the Loading indicator on opening a recording, but maybe the better solution is to show the loading only when eventsDataIsLoading && firstSnapshotChunkIsLoading. I'll make that change quickly

**Pull request merged**